decArgo_api/decoder_bindings/file_manager.py

```diff
@@ -0,0 +1,94 @@
+"""Code to prepare the input files to be decoded."""
+
+import logging
+import os
+import re
+import shutil
+import traceback
+from datetime import datetime
+from pathlib import Path
+
+from fastapi import UploadFile
+
+
+class FileManagerError(Exception):
+    """Raised when any error is encountered during the file management process."""
+
+
+class FileManager:
+    """Methods to prepare raw files for delivery to the Coriolis Decoder."""
+
+    def __init__(self, files: list[UploadFile]):
+        """Initialise the instance with the given files."""
+        self.files = files
+        self.base_archive_cycle_location = Path(
+            os.getenv("ARCHIVE_CYCLE_LOCATION", "/mnt/data/rsync/archive/cycle")
+        ).resolve()
+        self.base_archive_rsync_location = Path(
+            os.getenv("ARCHIVE_RSYNC_LOCATION", "/mnt/data/rsync/rsync_list/")
+        ).resolve()
+
+    def _detect_imei(self):
+        imei_number_search = re.search(r"\d{15}", self.files[0].filename)
+        if imei_number_search is None:
+            raise ValueError("No IMEI detected in filename!")
+        return imei_number_search.group()
+
+    def copy_file_to_input_directory(self, file) -> bool:
+        """Copy the file to a directory where the decoder can pick it up.
+
+        The purpose of returning True/False is for the next step in the chain which is to
+        produce the 'rsync' file.
+
+        If a copy is successful, the function returns True meaning the name will be written to the rsync file and
+        is therefore given to the decoder. If False, then the copy has failed, and will be logged accordingly.
+
+        Args:
+            file: The file to be copied.
+
+        Returns:
+            True if copy was successful, otherwise False.
+        """
+        float_directory = self.base_archive_cycle_location / self.imei
+        float_directory.mkdir(parents=True, exist_ok=True)  # Ensure the directory exists.
+        new_float_filename = float_directory / file.filename
+
+        try:
+            with new_float_filename.open("wb") as buffer:
+                shutil.copyfileobj(file.file, buffer)
+        except (OSError, TypeError):
+            logging.error("An error occurred during the file management process: %s", traceback.format_exc())
+            return False
+        return True
```

Comment:
What do we gain from the True/False returns? Would we be better off failing early if something goes wrong?

Reply:
This seems like a sharp edge to me... Would failing fast be better? Genuinely not sure – would like to consider opinions.